### PR TITLE
extract_shape: support for subregions in shapefiles

### LINKED ISSUE #273: add support for sub shapes to extract_shape
**Is your feature request related to a problem? Please describe.**

I am using extract_shape in my recipe to get the area averages for a given region as defined by a shapefile. eventually I need a different value for each sub shape in the shapefile. This is currently only possible by doing some sort of loop over different files. The work-in-progress PR adds a flag, decomposed, to do this directly

**Would you be able to help out?**
#272

### PR DESCRIPTION
**Pull request by ipelupessy**
preprocessor task

the resulting cube has an extra catchment_id dimension.

* * *

name: Core contribution
about:
title: ''
labels: ''
assignees: ''

* * *

Before you start, please read [CONTRIBUTING.md](https://github.com/ESMValGroup/ESMValTool/blob/version2_development/CONTRIBUTING.md).

**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] Add unit tests
-   [x] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [x] If writing a new/modified preprocessor function, please update the [documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/esmvalcore/preprocessor.html)
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   n/a Please use `yamllint` to check that your YAML files do not contain mistakes
-   n/a If you make backward incompatible changes to the recipe format, make a new pull request in the [ESMValTool repository](https://github.com/ESMValGroup/ESMValTool) and add the link below

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes #273

### CODE REVIEW

File: esmvalcore/preprocessor/_area.py [hunk 1]
```diff
@@ -357,15 +360,33 @@ def extract_shape(cube, shapefile, method='contains', crop=True):
         if lon_coord.ndim == 1 and lat_coord.ndim == 1:
             lon, lat = np.meshgrid(lon.flat, lat.flat, copy=False)
 
-        selection = np.zeros(lat.shape, dtype=bool)
+        if decomposed:
+            cubelist=iris.cube.CubeList()
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_area.py]: suggested change: `cubelist = iris.cube.CubeList()`

File: esmvalcore/preprocessor/_area.py [hunk 2]
```diff
@@ -374,25 +366,83 @@ def extract_shape(cube, shapefile, method='contains', crop=True,
...
+            lon, lat = np.meshgrid(lon.flat, lat.flat, copy=False)
+
+        cubelist = iris.cube.CubeList()
+
+        for item in geometries:
+            shape = shapely.geometry.shape(item['geometry'])
+            if method == 'contains':
+                select = shapely.vectorized.contains(shape, lon, lat)
+                if method == 'representative' or not select.any():
+                    select = _select_representative_point(shape, lon, lat)
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_area.py]: shouldn't be at the same indentation point as `if method == 'contains':`?

**Comment by JaroCamphuijsen** [comment]
@valeriupredoi I split it up in two as you requested.

**Comment by valeriupredoi** [comment]
@bouweandela have a look at the failed test, it appears that the CircleCI is not finding the `CODACY_PROJECT_TOKEN` variable that is set at project creation - since I haven't done this, can you have a go at fixing it plss? :beer:

**Comment by valeriupredoi** [comment]
ah that explains it, it was a fork - that confused both me and CodaCI

**Comment by bouweandela** [comment]
@JaroCamphuijsen or @ipelupessy could you please merge the `development` branch into this branch to check that the issue with circleci is indeed solved?

**Comment by ipelupessy** [comment]
@bouweandela do we need a branch in the ESMValCore repo for circleci? (its currently branch in my fork - I cannot push to ESMValCore..)

**Comment by bouweandela** [comment]
> @bouweandela do we need a branch in the ESMValCore repo for circleci? (its currently branch in my fork - I cannot push to ESMValCore..)

No, that should be fixed, now, that's why I asked you to merge the esmvalgroup/esmvalcore development branch, so you get the latest circleci configuration file.

**Comment by ipelupessy** [comment]
ok, done

**Comment by bouweandela** [comment]
CircleCI is working as expected now.

File: esmvalcore/preprocessor/_area.py [hunk 3]
```diff
@@ -357,15 +362,37 @@ def extract_shape(cube, shapefile, method='contains', crop=True):
...
                 select = _select_representative_point(shape, lon, lat)
-            selection |= select
 
-        selection = da.broadcast_to(selection, cube.shape)
-        cube.data = da.ma.masked_where(~selection, cube.core_data())
-        return cube
+            if decomposed:
+                cc = cube.copy()
+                coord = iris.coords.AuxCoord(
+                    int(item['properties']['ID']),
```
Comment (bouweandela) [inline esmvalcore/preprocessor/_area.py]: Is 'ID' always present? I've seen cases where it's called 'id' as well, see e.g. the files in esmvaltool/diag_scripts/shapeselect/testdata.

File: esmvalcore/preprocessor/_area.py [hunk 4]
```diff
@@ -357,15 +362,37 @@ def extract_shape(cube, shapefile, method='contains', crop=True):
...
+                    units='no_unit',
+                    long_name="catchment_ID"
```
Comment (bouweandela) [inline esmvalcore/preprocessor/_area.py]: This is very specific to hydrological use cases. It would just use 'ID' or 'id'.

Comment (bouweandela) [reply esmvalcore/preprocessor/_area.py]: or maybe 'shape_id'

Note that this adds an extra dimension to the resulting variable that is not specified in the CMOR table. @mattiarighi This shoudn't be a problem, right? I believe we have more preprocessor functions that change the dimensions of the data.

Comment (mattiarighi) [reply esmvalcore/preprocessor/_area.py]: Yes, several preprocessors reduce the dimensionality of the data, so I guess it is ok having one that increases it. The question

File: esmvalcore/preprocessor/_area.py [hunk 5]
```diff
@@ -357,15 +362,37 @@ def extract_shape(cube, shapefile, method='contains', crop=True):
...
+                cc.add_aux_coord(coord)
+
+                select = da.broadcast_to(select, cc.shape)
+                cc.data = da.ma.masked_where(~select, cc.core_data())
+                cubelist.append(cc)
+            else:
+                selection |= select
+
+        if decomposed:
+            cube = cubelist.merge()[0]
```
Comment (bouweandela) [inline esmvalcore/preprocessor/_area.py]:
```suggestion
            cube = cubelist.merge_cube()
```
to avoid accidentally loosing part of your data

**Comment by ipelupessy** [comment]
It has documentation and tests now..

File: esmvalcore/preprocessor/_area.py [hunk 6]
```diff
@@ -330,6 +331,10 @@ def extract_shape(cube, shapefile, method='contains', crop=True):
     crop: bool, optional
         Crop the resulting cube using `extract_region()`. Note that data on
         irregular grids will not be cropped.
+    decomposed: bool, optional
+        whether or not to retain the sub shapes of the shapefile in the output.
```
Comment (bouweandela) [inline esmvalcore/preprocessor/_area.py]:
```suggestion
        Whether or not to retain the sub shapes of the shapefile in the output.
```

**Comment by ipelupessy** [comment]
I am not going to fix the codacy issues. For me the code the code is ok. Its very annoying that it reports all tests to be passed only for it to flag it red the next day, btw.

**Comment by ipelupessy** [comment]
test does not discriminate between ```|``` and ```&``` in line 344 of _area.py...I think the latter is correct but have to recheck...

**Comment by valeriupredoi** [comment]
@ipelupessy please be advised of https://github.com/ESMValGroup/ESMValCore/pull/319 that is necessary when shapefiles are on a -180, +180deg grid (most of the cases I've encountered); also if you know how to check the grid in a shapefile, that'd be great since that way we can generalize the method :beer:

**Comment by ipelupessy** [comment]
ok, I have added a test so that it will fail when selection logic is changed too wrong operator

**Comment by ipelupessy** [comment]
(basically a code branch was not tested)

**Comment by ipelupessy** [comment]
> @ipelupessy please be advised of #319 that is necessary when shapefiles are on a -180, +180deg grid (most of the cases I've encountered); also if you know how to check the grid in a shapefile, that'd be great since that way we can generalize the method beer

ok, I think that change is compatible (can be added to _get_masks_from_geometries or just before the call to this).. for the shapefiles I think its matter of checking the range of the longitudes (but I am no expert on shapefiles)..in any case maybe it is a good idea to add a test case for this...

File: esmvalcore/preprocessor/_area.py [hunk 7]
```diff
@@ -346,31 +386,39 @@ def extract_shape(cube, shapefile, method='contains', crop=True):
...
-        raise ValueError(
-            "Invalid value for `method`. Choose from 'contains', ",
-            "'representative'.")
 
-    with fiona.open(shapefile) as geometries:
-        if crop:
-            cube = _crop_cube(cube, *geometries.bounds)
-        lon_coord = cube.coord(axis='X')
-        lat_coord = cube.coord(axis='Y')
+    geometries = fiona.open(shapefile)
```
Comment (bouweandela) [inline esmvalcore/preprocessor/_area.py]: Can you please restore the `with` block? That will ensure the shapefile is closed even if an exception is raised before `geometries.close()` is reached.

File: esmvalcore/preprocessor/_area.py [hunk 8]
```diff
@@ -315,7 +315,43 @@ def _select_representative_point(shape, lon, lat):
...
+        elif 'id' in item['properties']:
+            id_ = int(item['properties']['id'])
+        else:
+            id_ = i
+
+        selections[id_] = select
+
+    if not decomposed and len(selections) > 1:
+        selection = np.zeros(lat.shape, dtype=bool)
+        for id_, select in selections.items():
```
Comment (bouweandela) [inline esmvalcore/preprocessor/_area.py]:
```suggestion
        for select in selections.values():
```

**Comment by bouweandela** [comment]
Apart from the two minor comments above, this looks great! Could you address those?

**Comment by valeriupredoi** [comment]
> > @ipelupessy please be advised of #319 that is necessary when shapefiles are on a -180, +180deg grid (most of the cases I've encountered); also if you know how to check the grid in a shapefile, that'd be great since that way we can generalize the method beer
> 
> ok, I think that change is compatible (can be added to _get_masks_from_geometries or just before the call to this).. for the shapefiles I think its matter of checking the range of the longitudes (but I am no expert on shapefiles)..in any case maybe it is a good idea to add a test case for this...

cheers @ipelupessy -> that is needed indeed otherwise there are serious shifts in regions wrt actual desired position. I will have to find a way to check the coordinate inside the shapefile, but for now all the shapefiles Ive seen are all on -180/+180 deg grids, so it's probably safer to plug the shift in now rather than leave it for later :beer:

**Comment by bouweandela** [comment]
@valeriupredoi Let's fix the thing with the domain of the longitude coordinate in shapefiles in the other pull request, because I think this one is almost ready for merging.

**Comment by valeriupredoi** [comment]
> @valeriupredoi Let's fix the thing with the domain of the longitude coordinate in shapefiles in the other pull request, because I think this one is almost ready for merging.

sure! but mustn't forget about that :beer:

File: esmvalcore/preprocessor/_area.py (hunk 7 above)
Comment (ipelupessy) [reply esmvalcore/preprocessor/_area.py]: done

**Comment by ipelupessy** [comment]
@bouweandela who should merge this?

**Comment by bouweandela** [comment]
@mattiarighi Could you please have a look at this pull request?

**Comment by mattiarighi** [comment]
Looks good to me, but I have no test case to try it.

**Comment by bouweandela** [comment]
You can use the example files here: https://github.com/ESMValGroup/ESMValTool/tree/version2_development/esmvaltool/diag_scripts/shapeselect/testdata

multicatchment.shp contains multiple shapes, while the other files contain a single shape. Make sure to regrid to a small grid (e.g. 0.25x0.25 degrees), or you'll just select a single point for each catchment.

**Comment by mattiarighi** [comment]
> I think we may have preprocessor functions and diagnostics which assume that will be time.

Yes.

> Would you have a suggestion for a better location for this coordinate? I believe lat/lon are almost always the last two dimensions?

Yes, but adding it add the end might be less critical...

Are there examples in the CMOR tables for non-standard coordinates?

**Comment by bouweandela** [comment]
I found a pretty exotic coordinate called `dbze`, used e.g. in variable `cfadDbze94` from mip table `Emon`.

However, my impression from looking through the tables was that longitude and latitude are pretty consistently listed as the first two coordinates (so the last two of the cube), but the vertical and time axis are swapped in many cases, so I think that assuming that the time coordinate is the first dimension is very unsafe. @zklaus Do you know if there are any standards or recommendations for this?

**Comment by zklaus** [comment]
The order follows from [CF-1.7, 2.4](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#dimensions) and should always finish with (time, z, y, x). Any other dimension should preceed them, so I think the position is ok.

However, upon double checking the cmip6 tables I see that for the largest part they place time in the extreme position, even when there are other variables, but this is not always the case.
The complete list of counter-examples follows, but take it with a grain of salt because it seems in most cases the extraordinary coordinate is a scalar coordinate. I have to go to a meeting now, will investigate further after.

### Table 3hr
- huss: longitude latitude time1 height2m
- mrsos: longitude latitude time1 sdepth1
- tas: longitude latitude time1 height2m
- uas: longitude latitude time1 height10m
- vas: longitude latitude time1 height10m
### Table 6hrLev
- bs550aer: longitude latitude time1 lambda550nm scatter180
- ec550aer: longitude latitude alevel time1 lambda550nm
### Table 6hrPlev
- hurs: longitude latitude time height2m
- rv850: longitude latitude time p850
- sfcWind: longitude latitude time height10m
- tas: longitude latitude time height2m
- ua100m: longitude latitude time height100m
- uas: longitude latitude time height10m
- va100m: longitude latitude time height100m
- vas: longitude latitude time height10m
- wsgmax100m: longitude latitude time height100m
- wsgmax10m: longitude latitude time height10m
- zg1000: longitude latitude time p1000
### Table 6hrPlevPt
- huss: longitude latitude time1 height2m
- mrsos: longitude latitude time1 sdepth1
- rv850: longitude latitude time1 p850
- sfcWind: longitude latitude time1 height10m
- tas: longitude latitude time1 height2m
- tsl: longitude latitude time1 sdepth1
- uas: longitude latitude time1 height10m
- vas: longitude latitude time1 height10m
- vortmean: longitude latitude time1 pl700
- zg500: longitude latitude time1 p500
### Table AERday
- od550aer: longitude latitude time lambda550nm
- ua10: longitude latitude time p10
- zg10: longitude latitude time p10
- zg100: longitude latitude time p100
- zg1000: longitude latitude time p1000
- zg500: longitude latitude time p500
### Table AERhr
- tas: longitude latitude time height2m
### Table AERmon
- abs550aer: longitude latitude time lambda550nm
- od550aer: longitude latitude time lambda550nm
- od550aerh2o: longitude latitude time lambda550nm
- od550bb: longitude latitude time lambda550nm
- od550bc: longitude latitude time lambda550nm
- od550csaer: longitude latitude time lambda550nm
- od550dust: longitude latitude time lambda550nm
- od550lt1aer: longitude latitude time lambda550nm
- od550no3: longitude latitude time lambda550nm
- od550oa: longitude latitude time lambda550nm
- od550so4: longitude latitude time lambda550nm
- od550soa: longitude latitude time lambda550nm
- od550ss: longitude latitude time lambda550nm
### Table AERmonZ
- vt100: latitude time p100
### Table Amon
- hurs: longitude latitude time height2m
- huss: longitude latitude time height2m
- sfcWind: longitude latitude time height10m
- tas: longitude latitude time height2m
- tasmax: longitude latitude time height2m
- tasmin: longitude latitude time height2m
- uas: longitude latitude time height10m
- vas: longitude latitude time height10m
### Table CF3hr
- hurs: longitude latitude time1 height2m
- sfcWind: longitude latitude time1 height10m
### Table CFday
- clhcalipso: longitude latitude time p220
- cllcalipso: longitude latitude time p840
- clmcalipso: longitude latitude time p560
- ta700: longitude latitude time p700
- wap500: longitude latitude time p500
### Table CFmon
- clhcalipso: longitude latitude time p220
- cllcalipso: longitude latitude time p840
- clmcalipso: longitude latitude time p560
### Table CFsubhr
- hurs: site time1 height2m
- huss: site time1 height2m
- latitude: site
- longitude: site
- sfcWind: site time1 height10m
- tas: site time1 height2m
- uas: site time1 height10m
- vas: site time1 height10m
### Table day
- hurs: longitude latitude time height2m
- hursmax: longitude latitude time height2m
- hursmin: longitude latitude time height2m
- huss: longitude latitude time height2m
- mrsos: longitude latitude time sdepth1
- sfcWind: longitude latitude time height10m
- sfcWindmax: longitude latitude time height10m
- tas: longitude latitude time height2m
- tasmax: longitude latitude time height2m
- tasmin: longitude latitude time height2m
- uas: longitude latitude time height10m
- vas: longitude latitude time height10m
### Table E1hrClimMon
### Table E1hr
### Table E3hr
- sfcWind: longitude latitude time height10m
- uas: longitude latitude time height10m
- vas: longitude latitude time height10m
### Table E3hrPt
- clhcalipso: longitude latitude time1 p220
- cllcalipso: longitude latitude time1 p840
- clmcalipso: longitude latitude time1 p560
### Table E6hrZ
### Table Eday
- aod550volso4: longitude latitude time lambda550nm
- dmlt: longitude latitude time stempzero
- hursminCrop: longitude latitude time height2m
- hus850: longitude latitude time p850
- ta500: longitude latitude time p500
- ta850: longitude latitude time p850
- tasmaxCrop: longitude latitude time height2m
- tasminCrop: longitude latitude time height2m
### Table EdayZ
### Table Efx
- clayfrac: longitude latitude sdepth
- fldcapacity: longitude latitude sdepth
- ksat: longitude latitude sdepth
- ps: longitude latitude
- rld: alevhalf spectband
- rlu: alevhalf spectband
- rootdsl: longitude latitude sdepth
- rsd: alevhalf spectband
- rsu: alevhalf spectband
- sandfrac: longitude latitude sdepth
- sftflf: longitude latitude
- sftgrf: longitude latitude
- siltfrac: longitude latitude sdepth
- slthick: longitude latitude sdepth
- vegHeight: longitude latitude
- wilt: longitude latitude sdepth
### Table Emon
- cSoilAbove1m: longitude latitude time sdepth10
- cropFracC3: longitude latitude time typec3crop
- cropFracC4: longitude latitude time typec4crop
- ec550aer: longitude latitude alevel time lambda550nm
- grassFracC3: longitude latitude time typec3natg
- grassFracC4: longitude latitude time typec4natg
- hursminCrop: longitude latitude time height2m
- hussLut: longitude latitude landUse time height2m
- mrsosLut: longitude latitude landUse time sdepth1
- nwdFracLut: longitude latitude landUse time typenwd
- od550aerso: longitude latitude time lambda550nm
- od550so4so: longitude latitude time lambda550nm
- pastureFracC3: longitude latitude time typec3pastures
- pastureFracC4: longitude latitude time typec4pastures
- sfcWindmax: longitude latitude time height10m
- tasLut: longitude latitude landUse time height2m
- tasmaxCrop: longitude latitude time height2m
- tasminCrop: longitude latitude time height2m
- thetaot2000: longitude latitude time depth2000m
- thetaot300: longitude latitude time depth300m
- thetaot700: longitude latitude time depth700m
- treeFracBdlDcd: longitude latitude time typetreebd
- treeFracBdlEvg: longitude latitude time typetreebe
- treeFracNdlDcd: longitude latitude time typetreend
- treeFracNdlEvg: longitude latitude time typetreene
- vegFrac: longitude latitude time typeveg
- wetlandFrac: longitude latitude time typewetla
### Table EmonZ
### Table Esubhr
- huss: longitude latitude time1 height2m
- tas: longitude latitude time1 height2m
### Table Eyr
- baresoilFrac: longitude latitude time typebare
- cropFrac: longitude latitude time typecrop
- grassFrac: longitude latitude time typenatgr
- residualFrac: longitude latitude time typeresidual
- shrubFrac: longitude latitude time typeshrub
- treeFrac: longitude latitude time typetree
- vegFrac: longitude latitude time typeveg
### Table fx
- areacella: longitude latitude
- areacellr: longitude latitude
- mrsofc: longitude latitude
- orog: longitude latitude
- rootd: longitude latitude
- sftgif: longitude latitude
- sftlf: longitude latitude
- zfull: longitude latitude alevel
### Table grids
- latitude: longitude latitude
- longitude: longitude latitude
- vertices_latitude: vertices longitude latitude
- vertices_longitude: vertices longitude latitude
### Table IfxAnt
- areacellg: longitude latitude
- hfgeoubed: xant yant
- lithk: xant yant
- topg: xant yant
### Table IfxGre
- areacellg: longitude latitude
- hfgeoubed: xgre ygre
- lithk: xgre ygre
- topg: xgre ygre
### Table ImonAnt
- tas: time height2m
### Table ImonGre
- tas: time height2m
### Table IyrAnt
### Table IyrGre
### Table LImon
### Table Lmon
- baresoilFrac: longitude latitude time typebare
- burntFractionAll: longitude latitude time typeburnt
- c3PftFrac: longitude latitude time typec3pft
- c4PftFrac: longitude latitude time typec4pft
- cropFrac: longitude latitude time typecrop
- grassFrac: longitude latitude time typenatgr
- mrsos: longitude latitude time sdepth1
- pastureFrac: longitude latitude time typepasture
- residualFrac: longitude latitude time typeresidual
- shrubFrac: longitude latitude time typeshrub
- treeFrac: longitude latitude time typetree
- treeFracPrimDec: longitude latitude time typepdec
- treeFracPrimEver: longitude latitude time typepever
- treeFracSecDec: longitude latitude time typesdec
- treeFracSecEver: longitude latitude time typesever
### Table Oclim
### Table Oday
### Table Odec
### Table Ofx
- areacello: longitude latitude
- basin: longitude latitude
- deptho: longitude latitude
- hfgeou: longitude latitude
- masscello: longitude latitude olevel
- sftof: longitude latitude
- thkcello: longitude latitude olevel
- ugrido: longitude latitude
- volcello: longitude latitude olevel
### Table Omon
- dpco2: longitude latitude time depth0m
- dpco2abio: longitude latitude time depth0m
- dpco2nat: longitude latitude time depth0m
- dpo2: longitude latitude time depth0m
- eparag100: longitude latitude time depth100m
- epc100: longitude latitude time depth100m
- epcalc100: longitude latitude time depth100m
- epfe100: longitude latitude time depth100m
- epn100: longitude latitude time depth100m
- epp100: longitude latitude time depth100m
- epsi100: longitude latitude time depth100m
- fbddtalk: longitude latitude time olayer100m
- fbddtdic: longitude latitude time olayer100m
- fbddtdife: longitude latitude time olayer100m
- fbddtdin: longitude latitude time olayer100m
- fbddtdip: longitude latitude time olayer100m
- fbddtdisi: longitude latitude time olayer100m
- fddtalk: longitude latitude time olayer100m
- fddtdic: longitude latitude time olayer100m
- fddtdife: longitude latitude time olayer100m
- fddtdin: longitude latitude time olayer100m
- fddtdip: longitude latitude time olayer100m
- fddtdisi: longitude latitude time olayer100m
- fg13co2: longitude latitude time depth0m
- fg14co2abio: longitude latitude time depth0m
- fgco2: longitude latitude time depth0m
- fgco2abio: longitude latitude time depth0m
- fgco2nat: longitude latitude time depth0m
- fgdms: longitude latitude time depth0m
- fgo2: longitude latitude time depth0m
- fsfe: longitude latitude time depth0m
- fsn: longitude latitude time depth0m
- icfriver: longitude latitude time depth0m
- ocfriver: longitude latitude time depth0m
- ppos: longitude latitude time depth0m
- spco2: longitude latitude time depth0m
- spco2abio: longitude latitude time depth0m
- spco2nat: longitude latitude time depth0m
### Table Oyr
### Table SIday
- siconc: longitude latitude time typesi
- siconca: longitude latitude time typesi
### Table SImon
- siconc: longitude latitude time typesi
- siconca: longitude latitude time typesi
- simpconc: longitude latitude time typemp
- sirdgconc: longitude latitude time typesirdg

**Comment by zklaus** [comment]
It seems here CF and CMIP6 are at odds, but I guess the justification is about chunking ultimately. Anyways, probably it is better to stick to CMIP6 and put `shape_id` between `time` and `latitude`.

**Comment by bouweandela** [comment]
@ipelupessy Can you please implement the suggestion above? I.e. currently the order of dimensions of the resulting cube is: `shape_id [time] [other coordinates] latitude longitude` and it should be `[time] shape_id [other_coordinates] latitude longitude`, where the coordinates in square brackets are optional. You can probably use the function [`iris.cube.Cube.transpose`](https://scitools.org.uk/iris/docs/latest/iris/iris/cube.html#iris.cube.Cube.transpose) to move the time coordinate (if present) to the first position.

**Comment by ipelupessy** [comment]
will do..not today though..

**Comment by ipelupessy** [comment]
I have added a function fixing the order of dimensions; however I may be taking the wrong approach, since when I tried to add latitude and longitude in the smae manner I ran into the problem that:
```
lat_dim = cube.coord_dims('latitude')
lon_dim = cube.coord_dims('longitude')
```
can return the same tuple (e.g. (0,1))

**Comment by zklaus** [comment]
@ipelupessy this will happen when latitude and longitude are two-dimensional fields, for example in irregular grids as commonly used in the ocean or in projections from regional models.

That is why CF speaks of y and x in the order which may be (latitude, longitude), (grid_latitude, grid_longitude), (projection_y_coordinate, projection_x_coordinate), ... The point is that the 2 horizontal coordinates should be last.

**Pull request merged**